lib/bcrypt/password.rb

```diff
@@ -61,9 +61,14 @@ def initialize(raw_hash)
       end
     end
 
-    # Compares a potential secret against the hash. Returns true if the secret is the original secret, false otherwise.
+    # Compares a potential secret against the hash. Returns true if the secret
+    # is the original secret, false otherwise.
     def ==(secret)
-      super(BCrypt::Engine.hash_secret(secret, @salt))
+      if valid_hash?(secret)
+        super(secret)
```

Comment:
Potential edge case to consider: what if a user has a password that looks like a bcrypt hash?